verification/io/io-spec.gobra

```diff
@@ -38,75 +38,221 @@ type Place int
 
 ghost
+decreases
 pure func undefined() IO_dp3s_state_local
 
-pred dp3s_iospec_bio3s_enter(s IO_dp3s_state_local, t Place)
-/*
-∀⋆ (λv. case v of
-Internal_val1 pkt recvif newpkt nextif ⇒
-if pkt ∈ ibuf s (Some recvif) ∧ (∃currseg traversedseg.
-(∃hf1 fut. dp2_enter_guard pkt currseg traversedseg asid hf1 recvif fut) ∧
-dp3s_forward (Packet2 traversedseg (LeftSeg pkt) (MidSeg pkt) (RightSeg pkt)) newpkt
-nextif)
-then ExT (λT'. ExV (λX. CBio (IN bio3s_enter) t (Internal_val1 pkt recvif newpkt nextif) X T'
-⋆ dp3s_iospec_ordered (dp3s_add_obuf s nextif newpkt) T')) else Bool True | _ ⇒ Bool True)
-*/
-
-pred dp3s_iospec_bio3s_xover_up2down(s IO_dp3s_state_local, t Place)
-/*
-∀⋆ (λv. case v of
-Internal_val1 pkt recvif newpkt nextif ⇒
-if ∃currseg. ¬ ConsDir currseg ∧ (∃nextseg. ConsDir nextseg ∧
-(∃traversedseg intermediatepkt hf1 hf2.
-xover_up2down2_link_type asid hf1 hf2 ∧
-(∃nextfut. dp3s_xover_common s pkt currseg nextseg traversedseg intermediatepkt hf1
-hf2 nextfut recvif newpkt nextif)))
-then ExT (λT'. ExV (λX. CBio (IN bio3s_xover_up2down) t (Internal_val1 pkt recvif newpkt
-nextif) X T'
-)
-⋆ dp3s_iospec_ordered (dp3s_add_obuf s nextif newpkt) T')) else Bool True
-| _ ⇒ Bool True)
-*/
-
-pred dp3s_iospec_bio3s_xover_core(s IO_dp3s_state_local, t Place)
-/* s t ≡ ∀⋆ (λv. case v of
-Internal_val1 pkt recvif newpkt nextif ⇒
-if asid ∈ core ∧ (∃currseg nextseg.
-ConsDir nextseg = ConsDir currseg ∧ (∃traversedseg intermediatepkt hf1 hf2.
-xover_core2_link_type hf1 hf2 asid (ConsDir currseg) ∧
-(∃nextfut. dp3s_xover_common s pkt currseg nextseg traversedseg intermediatepkt hf1 hf2 nextfut recvif newpkt nextif)))
-then ExT (λT'. ExV (λX. CBio (IN bio3s_xover_core) t (Internal_val1 pkt recvif newpkt nextif) X
-T' * dp3s_iospec_ordered (dp3s_add_obuf s nextif newpkt) T')) else Bool True
-| _ ⇒ Bool True)
-*/
-
-pred dp3s_iospec_bio3s_exit(s IO_dp3s_state_local, t Place)
-/* s t ≡ ∀⋆ (λv. case v of
-Internal_val2 pkt newpkt nextif ⇒
-if pkt ∈ ibuf s None ∧ dp3s_forward_ext pkt newpkt nextif
-then ExT (λT'. ExV (λX. CBio (IN bio3s_exit) t (Internal_val2 pkt newpkt nextif) X T' ⋆ dp3s_iospec_ordered (dp3s_add_obuf s (Some nextif) newpkt) T'))
-else Bool True | _ ⇒ Bool True)
-*/
-
-pred CBioIO_bio3s_send(t Place)
+pred CBio_IN_bio3s_enter(t Place, v IO_val)
 
 ghost
-requires CBioIO_bio3s_send(t)
-pure func dp3s_iospec_bio3s_send_T(s IO_dp3s_state_local, t Place) Place
+requires CBio_IN_bio3s_enter(t, v)
+decreases
+pure func CBio_IN_bio3s_enter_T(s IO_dp3s_state_local, t Place, v IO_val) Place
 
+
+ghost
+requires len(currseg.Future) > 0
+decreases
+pure func dp3s_iospec_bio3s_create_guard_traversedseg(currseg IO_seg3) IO_seg3 {
+	return let uinfo := !currseg.ConsDir ?
+	upd_uinfo(currseg.UInfo, currseg.Future[0]) :
+	currseg.UInfo in
+	IO_seg3_ {
+		AInfo: currseg.AInfo,
+			UInfo: uinfo,
+			ConsDir: currseg.ConsDir,
+			Peer: currseg.Peer,
+			Past: currseg.Past,
+			Future: currseg.Future,
+			History: currseg.History}
+}
```

Comment:
here, I would indent the lines after return 😅 (in general, I would indent all line breaks in pure functions)